File: java/plugins/com/voxgig/sekreto/plugins/Azuresecrets.java
// Azure Key Vault, as a voxgig/plugin definition.
//
// A PLUGIN, NOT PART OF THE CORE: it opens a socket, and the core reads
// at most a local file. A chain that does not name this kind must not
// carry it, so the calling project imports this definition and hands it
// to Sekreto in the `plugins` option.
// See docs/design/plugin-providers.md.

package com.voxgig.sekreto.plugins;

import static com.voxgig.sekreto.Addr.checkaddr;
import static com.voxgig.sekreto.plugins.Httpjson.dig;
import static com.voxgig.sekreto.plugins.Httpjson.fetchjson;
import static com.voxgig.sekreto.plugins.Httpjson.first;
import static com.voxgig.sekreto.plugins.Httpjson.headers;
import static com.voxgig.sekreto.plugins.Httpjson.renewtime;
import static com.voxgig.sekreto.plugins.Httpjson.trimslash;

import com.voxgig.sekreto.Provider;
import com.voxgig.sekreto.Sekreto;
import com.voxgig.sekreto.Sekreto.SekretoError;
import com.voxgig.sekreto.Support;
import com.voxgig.sekreto.plugins.Httpjson.Answer;
import voxgig.plugin.Definition;

/**
 * Azure Key Vault.
 *
 * <p>`api.token` reads secret `api-token` (dots flattened to `-`; Key
 * Vault names allow nothing else), current version. The token comes from
 * config, then a client-credentials login when tenant/clientid/
 * clientsecret are given, then the IMDS managed-identity endpoint - so on
 * Azure's own platform no credential configuration is needed.
 *
 * <p>As with GCP, the IMDS call is plain http to a link-local host by
 * platform design and carries no credential; the login and vault
 * addresses are `checkaddr`-guarded.
 */
public final class Azuresecrets implements Provider {

  /** The `azuresecrets` kind: what `plugins` hands to Sekreto. */
  public static final Definition PLUGIN = Support.providerplugin("azuresecrets", spec ->
      new Azuresecrets(
          Support.text(spec.get("vault")), Support.text(spec.get("token")),
          Support.text(spec.get("tenant")), Support.text(spec.get("clientid")),
          Support.text(spec.get("clientsecret")), Support.text(spec.get("loginaddr")),
          Support.text(spec.get("imdsaddr")), Support.text(spec.get("apiversion"))));
  private static final String RESOURCE = "https://vault.azure.net";

  private final String vault;
  private final String token;
  private final String tenant;
  private final String clientid;
  private final String clientsecret;
  private final String loginaddr;
  private final String imdsaddr;
  private final String apiversion;

  // A configured token is kept forever; logged-in and IMDS tokens carry
  // expires_in and are renewed shortly before they run out.
  private String livetoken;
  private long renewat = Long.MAX_VALUE;

  public Azuresecrets(String vault, String token, String tenant, String clientid,
      String clientsecret, String loginaddr, String imdsaddr, String apiversion) {
    this.vault = vault;
    this.token = token;
    this.tenant = tenant;
    this.clientid = clientid;
    this.clientsecret = clientsecret;
    this.loginaddr = loginaddr;
    this.imdsaddr = imdsaddr;
    this.apiversion = apiversion;
  }

  private String login() {
    if (null != token && !token.isEmpty()) {
      return token;
    }

    if (null != tenant && !tenant.isEmpty()
        && null != clientid && !clientid.isEmpty()
        && null != clientsecret && !clientsecret.isEmpty()) {
      String useloginaddr = first(loginaddr, "https://login.microsoftonline.com");
      checkaddr(useloginaddr);

      String url = trimslash(useloginaddr) + "/" + tenant + "/oauth2/v2.0/token";
      String form = "grant_type=client_credentials&client_id=" + Sigv4.uriescape(clientid)
          + "&client_secret=" + Sigv4.uriescape(clientsecret)
          + "&scope=" + Sigv4.uriescape(RESOURCE + "/.default");

      Answer res = fetchjson(
          "POST", url, headers("content-type", "application/x-www-form-urlencoded"), form);

      Object got = dig(res.body, "access_token");
      if (200 != res.status || null == got || String.valueOf(got).isEmpty()) {
        throw new SekretoError("sekreto: azure login failed: " + res.status);
      }

      renewat = renewtime(dig(res.body, "expires_in"));
      return String.valueOf(got);
    }

    String imds = trimslash(first(imdsaddr, "http://169.254.169.254"))
        + "/metadata/identity/oauth2/token?api-version=2018-02-01&resource="
        + Sigv4.uriescape(RESOURCE);

    Answer res = fetchjson("GET", imds, headers("Metadata", "true"), null);

    Object got = dig(res.body, "access_token");
    if (200 != res.status || null == got || String.valueOf(got).isEmpty()) {
      throw new SekretoError(
          "sekreto: azure: no token, no client credentials, and IMDS did not answer");
    }

    renewat = renewtime(dig(res.body, "expires_in"));
    return String.valueOf(got);
  }

  @Override
  public String lookup(String name) {
    String usevault = null == vault ? "" : vault;
    if (usevault.isEmpty()) {
      throw new SekretoError("sekreto: azure: no vault");
    }

    // Only an explicit scheme is a URL; a vault NAMED httpvault must
    // still become https://httpvault.vault.azure.net.
    String vaulturl =
        usevault.startsWith("http://") || usevault.startsWith("https://")
            ? usevault
            : "https://" + usevault + ".vault.azure.net";
    checkaddr(vaulturl);

    if (null == livetoken || System.currentTimeMillis() >= renewat) {
      livetoken = login();
    }

    String url = trimslash(vaulturl) + "/secrets/" + Sekreto.flatname(name, "-")
        + "?api-version=" + first(apiversion, "7.4");

    Answer res = fetchjson("GET", url, headers("authorization", "Bearer " + livetoken), null);

    if (404 == res.status) {
      return null;
    }

    if (200 != res.status) {
      throw new SekretoError(
          "sekreto: azure error: " + res.status + ": " + url.split("\\?")[0]);
    }

    Object value = dig(res.body, "value");
    return null == value ? null : String.valueOf(value);
  }

  @Override
  public String describe() {
    return "azuresecrets:" + (null == vault ? "" : vault);
  }
}

File: lua/src/sekreto/plugins/azuresecrets.lua
-- The azuresecrets plugin: Azure Key Vault, over HTTPS, with a
-- client-credentials login and the IMDS managed-identity endpoint.
--
-- A port of typescript/plugins/azuresecrets.ts, which is canonical.

local err = require('sekreto.err')
local name = require('sekreto.name')
local addr = require('sekreto.addr')
local providers = require('sekreto.providers')
local json = require('sekreto.plugins.json')
local httpjson = require('sekreto.plugins.httpjson')
local support = require('sekreto.plugins.support')

local fail = err.fail
local nonempty = providers.nonempty
local trimslash = providers.trimslash
local providerplugin = providers.providerplugin
local first = support.first
local nowms = support.nowms
local renewtime = support.renewtime
local NEVER = support.NEVER
local fetchjson = httpjson.fetchjson
local uriescape = support.uriescape

local M = {}

local AZURERESOURCE = 'https://vault.azure.net'

--- Azure Key Vault.
---
--- `api.token` reads secret `api-token` (dots flattened to `-`; Key Vault
--- names allow nothing else), current version. The token comes from
--- config, then a client-credentials login when tenant/clientid/
--- clientsecret are given, then the IMDS managed-identity endpoint.
local function azuresecrets(vault, token, tenant, clientid, clientsecret,
  loginaddr, imdsaddr, apiversion)
  local livetoken = nil
  local renewat = NEVER

  local function login()
    if nonempty(token) then
      return token
    end

    if nonempty(tenant) and nonempty(clientid) and nonempty(clientsecret) then
      local useloginaddr = first(loginaddr, 'https://login.microsoftonline.com')
      addr.checkaddr(useloginaddr)

      local url = trimslash(useloginaddr) .. '/' .. tenant .. '/oauth2/v2.0/token'
      local form = 'grant_type=client_credentials&client_id=' ..
        uriescape(clientid) ..
        '&client_secret=' .. uriescape(clientsecret) ..
        '&scope=' .. uriescape(AZURERESOURCE .. '/.default')

      local res = fetchjson(
        'POST', url,
        { { 'content-type', 'application/x-www-form-urlencoded' } },
        form
      )

      local got = json.text(json.dig(res.body, 'access_token'))
      if 200 ~= res.status or nil == got or '' == got then
        fail('sekreto: azure login failed: ' .. res.status)
      end

      renewat = renewtime(json.dig(res.body, 'expires_in'))
      return got
    end

    local imds = trimslash(first(imdsaddr, 'http://169.254.169.254')) ..
      '/metadata/identity/oauth2/token?api-version=2018-02-01&resource=' ..
      uriescape(AZURERESOURCE)

    local res = fetchjson('GET', imds, { { 'Metadata', 'true' } })

    local got = json.text(json.dig(res.body, 'access_token'))
    if 200 ~= res.status or nil == got or '' == got then
      fail('sekreto: azure: no token, no client credentials, and IMDS did not answer')
    end

    -- IMDS sends expires_in as a STRING, unlike everyone else.
    renewat = renewtime(json.dig(res.body, 'expires_in'))
    return got
  end

  return {
    lookup = function(secret)
      local usevault = vault or ''
      if '' == usevault then
        fail('sekreto: azure: no vault')
      end

      -- Only an explicit scheme is a URL; a vault NAMED httpvault must
      -- still become https://httpvault.vault.azure.net.
      local vaulturl
      if 'http://' == usevault:sub(1, 7) or 'https://' == usevault:sub(1, 8) then
        vaulturl = usevault
      else
        vaulturl = 'https://' .. usevault .. '.vault.azure.net'
      end
      addr.checkaddr(vaulturl)

      if nil == livetoken or nowms() >= renewat then
        livetoken = login()
      end

      local url = trimslash(vaulturl) .. '/secrets/' .. name.flatname(secret, '-') ..
        '?api-version=' .. first(apiversion, '7.4')

      local res = fetchjson('GET', url, { { 'authorization', 'Bearer ' .. livetoken } })

      if 404 == res.status then
        return nil
      end

      if 200 ~= res.status then
        fail('sekreto: azure error: ' .. res.status .. ': ' .. addr.bare(url))
      end

      return json.text(json.dig(res.body, 'value'))
    end,
    describe = function()
      return 'azuresecrets:' .. (vault or '')
    end,
  }
end

--- The kind, as a voxgig/plugin definition.
M.azuresecrets = providerplugin('azuresecrets', function(spec)
  return azuresecrets(spec.vault, spec.token, spec.tenant, spec.clientid,
    spec.clientsecret, spec.loginaddr, spec.imdsaddr, spec.apiversion)
end)

return M
